src/Elmish/Dispatch.purs

```diff
@@ -1,45 +1,90 @@
+  , handleMaybe
+  )
+  where
 
 import Prelude
 
 import Data.Maybe (Maybe, maybe)
 import Effect (Effect)
-import Effect.Uncurried (EffectFn1, EffectFn2, mkEffectFn1, mkEffectFn2) as E
+import Effect.Uncurried as E
+import Safe.Coerce (coerce)
+import Type.Equality (class TypeEquals)
 
 -- | A function that a view can use to report messages originating from JS/DOM.
 type Dispatch msg = msg -> Effect Unit
 
 infixr 9 handle as <|
 infixr 9 handleMaybe as <?|
 
--- | A convenience function to make construction of event handlers with
--- | arguments (i.e. `EffectFn1`) a bit shorter. The function takes a `Dispatch`
--- | and a mapping from the event argument to a message which the given
--- | `Dispatch` accepts, and it's also available in operator form.
--- |
--- | The following example demonstrates expected usage of both `handle` (in its
--- | operator form `<|`) and `handleMaybe` (in its operator form `<?|`):
--- |
--- |     textarea
--- |       { value: state.text
--- |       , onChange: dispatch <?| \e -> TextChanged <$> eventTargetValue e
--- |       , onMouseDown: dispatch <| \_ -> TextareaClicked
--- |       }
--- |
--- |       where
--- |         eventTargetValue = readForeign >=> lookup "target" >=> readForeign >=> lookup "value"
--- |
-handle :: forall arg msg. Dispatch msg -> (arg -> msg) -> E.EffectFn1 arg Unit
-handle dispatch fn = E.mkEffectFn1 $ dispatch <<< fn
-
--- | Same as `handle`, but dispatches a message optionally. See comments on
--- | `handle` for an example.
-handleMaybe :: forall arg msg. Dispatch msg -> (arg -> Maybe msg) -> E.EffectFn1 arg Unit
-handleMaybe dispatch fn = E.mkEffectFn1 $ maybe (pure unit) dispatch <<< fn
+class Handle msg event f where
+    -- | A convenience function to make construction of event handlers with
+    -- | arguments (i.e. `EffectFn1`) a bit shorter. The first parameter is a
+    -- | `Dispatch`. The second parameter can be either a message or a function
+    -- | from the event object to a message.
+    -- |
+    -- | Expected usage for this function is in its operator form `<|`
+    -- |
+    -- |     textarea
+    -- |       { value: state.text
+    -- |       , onChange: dispatch <| \e -> TextAreaChanged (E.textareaText e)
+    -- |       , onMouseDown: dispatch <| TextareaClicked
+    -- |       }
+    -- |
+    handle :: Dispatch msg -> f -> E.EffectFn1 event Unit
+
+class HandleMaybe msg event f where
+    -- | A variant of `handle` (aka `<|`) that allows to dispatch a message or
+    -- | not conditionally via returning a `Maybe message`.
+    -- |
+    -- | Expected usage for this function is in its operator form `<?|`
+    -- |
+    -- |     div
+    -- |       { onMouseDown: dispatch <?| \(E.MouseEvent e) ->
+    -- |           if e.ctrlKey then Just ClickedWithControl else Nothing
+    -- |       }
+    -- |
+    handleMaybe :: Dispatch msg -> f -> E.EffectFn1 event Unit
+
+instance (TypeEquals msg output, TypeEquals event input) => Handle msg event (input -> output) where
```

Comment:
What’s the difference between this and:

```suggestion
instance Handle msg event (event -> msg) where
```

Reply:
Ohhhh, but it does!

Imagine, for example, `msg ~ Either Int String`. So it would look like this:

```
  dispatch :: Dispatch (Either Int String)

  onClick: dispatch <| \e -> Left 42
```

How would the compiler resolve which instance of `Handle` to use here? Think about it for a second. [Now check it out live](https://try.purescript.org/?code=LYewJgrgNgpgBAWQIYEsB2cDuALGAnGAKEJWAAcQ8AXOABQKgjCNIurgFEAzLmAYxoAKbrwEBKEuUo0AIkipIAdBxRVceOMNXrBixWImFgqDAC5TnHvxoBVNKqMm4AXjhkIBOBHtViVAJ5k8DIoAM5k8nzYcMChAOYuMfFwALQAfJaitj7EaDCYAUGcAG4waFQAEkhoYLAa8K4cFZrw6XB2qmKpKXBoIDSs0tVUADRwAJJwAFYQoTR5MGBwoSDA8IVEhHxQSKGhcFU1dUkJMKXlcFxY6vCEcHDY1bXw5nAh4ZHRsQltV20c50qT2OZzKvhIaDm1T48EOzw03zgoIugmRNDa3y6OHwRHujyO8AA%2BnBiY1mgAdYltbwOe6EMAoYpwV4AbzgIDQAGEoCg%2BABrZkWAFguHHcYXAC%2BqQyHV8DKZpK8OUIG0QMD2SDiDU42nwEwuAGUqHh0HFiMUUPlBW8whEqFE1RqtdL2jkLVaGR97dFXHc4PK-fc2Rzuby%2BRZ8c9-bbPnBySgXQAZGBcAaBuASv1AA)

See what happens? "_The instance head contains unknown type variables_." The compiler doesn't even start resolving the instance until it knows all types. Because, you know, it need to know the type of, for example, parameter `e` in order to check whether it matches `SyntheticEvent`. And similarly with `msg`.

So instead I have to make those instances match on _any_ types whatsoever, as long as the third one is a function, so that all the compiler needs to determine is that it's a function, and then the instance becomes known. And only then, once the instance is matched, I apply constraints fixing `event ~ input` and `output ~ msg`.